expression-language.md

```diff
@@ -0,0 +1,375 @@
+on SQL. Note that neither the SQL standard (ISO 9075) nor the JMS standard nor any other SQL dialect are used as a
+normative foundation or constrain the expression syntax defined in this specification, but the syntax is informed by
+them.
+
+CESQL is a _[Total pure functional programming language][total-programming-language-wiki]_ in order to guarantee the
+termination of the evaluation of the expression. It features a type system correlated to the [CloudEvents type
+system][ce-type-system] and it features boolean and arithmetics operations, as well as built-in functions for string
+manipulation.
+
+The language is not constrained to a particular execution environment, which means it might run in a source, in a
+producer, in an intermediary, and it can be implemented using any tech stack.
+
+The CloudEvents Expression Language assumes the input always includes, but it's not limited to, a single valid and type
+checked CloudEvent instance. An expression MUST NOT mutate the value of the input CloudEvent instance, nor any of the
+other input values. The evaluation of an expression observes the concept of [referential
+transparency][referential-transparency-wiki]. The output of a CESQL expression evaluation is always a boolean _true_ or _false_ and it might include an error.
+
+The CloudEvents Expression Language doesn't support the handling of the data field of the CloudEvent instances, due to
+its polymorphic nature and complexity. We strongly encourage users that needs this functionality to use other more
+appropriate tools.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and
+"OPTIONAL" in this document are to be interpreted as described in [RFC2119][rfc2119].
+
+### 1.2. Relation to the Subscriptions API
+
+The CESQL can be used as a [filter dialect][subscriptions-filter-dialect] to filter on the input values.
+
+<!-- TODO -->
+
+## 2. Language syntax
+
+The grammar of the language is defined using the EBNF Notation from [W3C XML specification][ebnf-xml-spec].
+
+Although in the EBNFs keywords are defined using uppercase characters, they are case-insensitive. For example:
+
+```
+int(hop) < int(ttl) and int(hop) < 1000
+```
+
+Has the same syntactical meaning of:
+
+```
+INT(hop) < INT(ttl) AND INT(hop) < 1000
+```
+
+### 2.1 Expression
+
+The root of the expression is the `expression` rule:
+
+```ebnf
+expression ::= value-identifier | literal | unary-operation | binary-operation | function-invocation | ( "(" expression ")" )
+```
+
+Nested expressions MUST be correctly parenthesized.
+
+### 2.2. Value identifiers and literals
+
+Value identifiers in CESQL have the same restrictions of the [Attribute Naming
+Convention][ce-attribute-naming-convention] from the CloudEvents spec. A value identifier MUST NOT be greater than 20
+characters in length.
+
+```ebnf
+lowercase-char ::= [a-z]
+value-identifier ::= lowercase-char ( lowercase-char | digit )*
```

Comment:
oddly enough, CE allows for names to start with a digit. I'm not sure if that is correct or not, but we should align with it.

Reply:
Ah ok I wasn't aware of that, let me fix it